File: pset6/credit/credit.py
#  this program prompts the user for a credit card number
#  reports whether it is a valid American Express, MasterCard or Visa card.


from cs50 import get_int
import sys


def main():  # main function
    usage()

    cardNumber = get_int("Number: ")  # prompt user for input
    validate(cardNumber)


def usage():
    msg = """

SYNOPSIS
        python3 {}
        Number: [CREDIT CARD NUMBER]

    """.format(sys.argv[0])

    print(msg)


def validate(cc):  # validate credit card number
    tempList = [int(c) for c in str(cc)]

    listOne = tempList[-2::-2]  # second-to-last digit of credit card
    listTwo = tempList[::-2]  # remaining digits

    totalSum = sum(listTwo)
    for digit in listOne:
        totalSum += sum(divmod(2 * digit, 10))

    valid = totalSum % 10 == 0

    # check credit card's brands
    if tempList[0] == 4 and len(tempList) >= 13 and len(tempList) <= 16 and valid:
        print("VISA")
    elif tempList[0:2:] == [5, 1] or tempList[0:2:] == [5, 5] and len(tempList) == 16 and valid:
        print("MASTERCARD")
    elif tempList[0:2:] == [3, 4] or tempList[0:2:] == [3, 7] and len(tempList) == 15 and valid:
        print("AMEX")
    else:
        print("INVALID")


main()

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

// this program prompts the user for a credit card number and then reports whether it is a valid American Express, MasterCard, Visa card number.

int main()
{
    // prompt the user for input
    long long askCardNum = get_long_long("Number: ");

    // delcare variables
    int digitOne = 0;
    int digitTwo = 0;
    int numOfDigits = 0;
    int sumOfDoubleOdds = 0;
    int sumOfEvens = 0;

    // On iteration this loop do the checksum and validate
    while (askCardNum > 0)
    {
        digitTwo = digitOne;
        digitOne = askCardNum % 10;

        if (numOfDigits % 2 == 0)
        {
            sumOfEvens = sumOfEvens + digitOne;
        }
        else 
        {
            int multiply = 2 * digitOne;
            sumOfDoubleOdds += (multiply / 10) + (multiply % 10);
        }
        askCardNum = askCardNum / 10;
        numOfDigits++;
    }


    // report for the user

  
    bool valid = (sumOfEvens + sumOfDoubleOdds) % 10 == 0;
    int firstTwoDigits = (digitOne * 10) + digitTwo;

    if (digitOne == 4 && numOfDigits >= 13 && numOfDigits <= 16 && valid)
    {
        printf("VISA\n");
    }
    else if (firstTwoDigits >= 51 && firstTwoDigits <= 55 && numOfDigits == 16 && valid)
    {
        printf("MASTERCARD\n");
    }
    else if ((firstTwoDigits == 34 || firstTwoDigits == 37) && numOfDigits == 15 && valid)
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
    
}
